### PR TITLE
feat(desktop): workspace number badges on modifier hold

### LINKED ISSUE #3217: [feat] Show workspace number badges when holding Cmd/Ctrl for shortcut discoverability
## Summary

Superset already supports `Cmd+0` through `Cmd+9` (Mac) / `Ctrl+0` through `Ctrl+9` (Windows/Linux) to switch between workspaces, but there's no visual indicator showing which number corresponds to which workspace. Adding number badges would be a great quality of life improvement for discoverability.

## Proposed Solution

When the user holds `Cmd` (Mac) or `Ctrl` (Windows/Linux), show numbered badges (0–9) on the first 10 workspaces in the sidebar, numbered top-to-bottom across all projects. The badges should:

- Appear on workspace rows, positioned to the left of (or temporarily replacing) the git diff counts (e.g. `+456 -0`)
- Only be visible while the modifier key is held (momentary overlay)
- Use a small badge/pill style that feels native to the existing sidebar UI
- Workspaces beyond the first 10 simply don't get a badge

### Handling collapsed projects

When a project is collapsed and contains workspaces that fall within the 0–9 range, there are two reasonable approaches:

**Option A — Show numbers on the collapsed project row**: Display the relevant numbers (e.g. `3, 4`) on the project row itself so users know what shortcuts are available. Pressing the shortcut switches to that workspace and reveals it. This keeps numbering stable regardless of collapse state.

**Option B — Skip collapsed workspaces**: Only number visible/expanded workspaces. Numbers shift as projects are expanded/collapsed, but every visible badge always corresponds to something on screen.

Both have tradeoffs — stable muscle memory (A) vs. WYSIWYG clarity (B).

### PR DESCRIPTION
**Pull request by tvrmsmith**
**Links**
- Closes https://github.com/superset-sh/superset/issues/3217
- Plan: `apps/desktop/plans/20260408-1200-workspace-number-badges.md`

## Summary
- Show numbered shortcut badges (⌘1–⌘9) on workspace sidebar rows when holding the modifier key, making keyboard shortcuts discoverable
- Add a "Show workspace numbers on modifier hold" toggle in Settings → Behavior
- Badges use platform-aware labels (⌘1 on Mac, Ctrl+Shift+1 on Windows/Linux) derived from actual hotkey bindings, respecting user overrides

## Why / Context
Users don't know they can press ⌘1–⌘9 to jump between workspaces. This feature surfaces those shortcuts on modifier hold, similar to how Arc browser reveals tab numbers.

## How It Works
- **Modifier key state store** (Zustand) tracks which modifier keys are currently held
- **`useWorkspaceShortcutModifiers`** hook parses JUMP_TO_WORKSPACE_1–9 bindings to extract modifier keys and generate platform-specific display labels via `formatHotkeyDisplay`
- **`useModifierKeyListener`** hook attaches keydown/keyup/blur listeners, computes `isModifierHeld` by checking held keys against the shortcut combos
- **`WorkspaceShortcutBadge`** presentational pill component used for both modifier-held and hover states
- When modifier is held, the badge replaces diff stats; on hover (without modifier), badge appears alongside the close button
- Setting stored in local SQLite via `showWorkspaceNumbersOnModifier` column, exposed through tRPC get/set procedures
- Collapsed project headers show aggregated badges for contained workspaces

### Modifier Hold Behavior
- [ ] Hold ⌘ (Mac) / Ctrl+Shift (Win/Linux) → badges appear on first 9 workspaces
- [ ] Release modifier → badges disappear, diff stats return
- [ ] Badges show correct platform symbol (⌘1 not Ctrl+Shift+1 on Mac)
- [ ] Badge sizing is consistent regardless of diff stats width
- [ ] Window blur clears all held keys

### Hover Behavior
- [ ] Hover workspace row (no modifier held) → shortcut badge + close button appear
- [ ] Hover badge matches modifier-held badge format

### Settings
- [ ] Settings → Behavior → "Show workspace numbers on modifier hold" toggle works
- [ ] Setting persists after app restart
- [ ] When disabled, holding modifier does nothing
- [ ] Setting is searchable in settings search

### Collapsed Project Headers
- [ ] Collapse a project section → hold modifier → badges appear on project header
- [ ] Badges show in both expanded and collapsed sidebar modes

### Edge Cases
- [ ] More than 9 workspaces → only first 9 get badges
- [ ] Workspaces in sections get correct sequential numbering
- [ ] Custom hotkey overrides reflected in badge labels

## Testing
- `bun test` — 9 tests pass (modifier key store + shortcut modifier parsing)
- `npx biome check` — clean
- Manual testing in dev mode

## New Files
- `stores/modifier-key-state.ts` + `.test.ts` — Zustand store for held modifier keys
- `hooks/useWorkspaceShortcutModifiers.ts` + `.test.ts` — Parses shortcut bindings
- `hooks/useModifierKeyListener.ts` — Side-effect hook for keydown/keyup
- `WorkspaceShortcutBadge.tsx` — Presentational badge component

---
## Summary by cubic
Show numbered workspace shortcut badges when holding the workspace modifier key to make jump shortcuts easy to discover. Adds an opt‑in Behavior setting; works in both `WorkspaceSidebar` (v1) and `DashboardSidebar` (v2), with hover hints and collapsed project header badges.

- New Features
  - Hold the modifier to reveal badges 1–9 using real bindings via `formatHotkeyDisplay`; respects platform and user overrides.
  - Settings → Behavior → “Show workspace numbers on modifier hold”; persisted via local DB + tRPC; mounts the key listener only when enabled in both sidebars.
  - `useModifierKeyListener` + a modifier-key Zustand store track held keys; `useWorkspaceShortcutModifiers` parses bindings; `WorkspaceShortcutBadge` renders the pill; collapsed project headers show aggregated badges.

- Bug Fixes
  - Adapt to the `ShortcutBinding` object form by extracting `chord`; update the hotkey recorder test to accept Alt+key on Mac and read `.codeChord`.
  - Render the modifier-held badge outside the diff‑stats grid; remove a dead prop; add the missing local‑db migration and mark the settings‑search item as “shared” so the toggle appears in v2.
  - Guard `process.env.SUPERSET_WORKSPACE_NAME` in production builds to avoid redirecting user data.

<sup>Written for commit 6a638c797fb6df1633e629105fc03fe3397efdd0. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/superset-sh/superset/pull/3269?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Toggle to show workspace numbers as badges while holding the shortcut modifier; setting persists.
  * Workspace shortcuts now render as dynamic numbered badges (with per-project labels and tooltips) when the modifier key is held.
* **Tests**
  * Added/updated tests covering modifier-key handling and hotkey parsing/recording.
* **Documentation**
  * Added planning docs describing the workspace-number badges feature and rollout steps.

### CODE REVIEW

File: apps/desktop/src/renderer/screens/main/components/WorkspaceSidebar/WorkspaceListItem/WorkspaceListItem.tsx [hunk 1]
```diff
@@ -388,46 +400,50 @@ export function WorkspaceListItem({
...
+														label="Close workspace"
+														id={isActive ? "CLOSE_WORKSPACE" : undefined}
+													/>
+												</TooltipContent>
+											</Tooltip>
+										)}
+									</div>
 								</div>
-							</div>
+							)}
```
[the remark this answers is not in the document]
Comment (TrevorSmith-Wellsky) [reply apps/desktop/src/renderer/screens/main/components/WorkspaceSidebar/WorkspaceListItem/WorkspaceListItem.tsx:453]: I already saw this and thought it was acceptable to not show the 'X' on hover when the modifier is pressed because it added weird spacing to the worktree workspaces.

File: apps/desktop/src/renderer/screens/main/components/WorkspaceSidebar/WorkspaceListItem/WorkspaceDiffStats.tsx [hunk 2]
```diff
@@ -4,18 +4,22 @@ interface WorkspaceDiffStatsProps {
 	additions: number;
 	deletions: number;
 	isActive?: boolean;
+	/** When true, hide the diff stats (e.g., when modifier key is held to show shortcut badges) */
+	hidden?: boolean;
 }
```
[the remark this answers is not in the document]
Comment (TrevorSmith-Wellsky) [reply apps/desktop/src/renderer/screens/main/components/WorkspaceSidebar/WorkspaceListItem/WorkspaceDiffStats.tsx:7]: Fixed — removed the dead `hidden` prop in 8896b8729.

File: apps/desktop/src/renderer/screens/main/components/WorkspaceSidebar/WorkspaceListItem/WorkspaceShortcutBadge.tsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import { cn } from "@superset/ui/utils";
+
+interface WorkspaceShortcutBadgeProps {
+	/** The key label to display (e.g., "1", "2") */
+	label: string;
```
[the remark this answers is not in the document]
Comment (TrevorSmith-Wellsky) [reply apps/desktop/src/renderer/screens/main/components/WorkspaceSidebar/WorkspaceListItem/WorkspaceShortcutBadge.tsx]: Fixed — updated the JSDoc to reflect the actual formatted shortcut labels in 8896b8729.

File: apps/desktop/src/renderer/screens/main/components/WorkspaceSidebar/WorkspaceListItem/WorkspaceShortcutBadge.tsx [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import { cn } from "@superset/ui/utils";
+
+interface WorkspaceShortcutBadgeProps {
+	/** The key label to display (e.g., "1", "2") */
```
[the remark this answers is not in the document]
Comment (TrevorSmith-Wellsky) [reply apps/desktop/src/renderer/screens/main/components/WorkspaceSidebar/WorkspaceListItem/WorkspaceShortcutBadge.tsx]: Fixed — updated the JSDoc to reflect the actual formatted shortcut labels in 8896b8729.

File: packages/local-db/src/schema/schema.ts [hunk 5]
```diff
@@ -222,6 +222,10 @@ export const settings = sqliteTable("settings", {
 	worktreeBaseDir: text("worktree_base_dir"),
 	openLinksInApp: integer("open_links_in_app", { mode: "boolean" }),
 	defaultEditor: text("default_editor").$type<ExternalApp>(),
+	showWorkspaceNumbersOnModifier: integer(
+		"show_workspace_numbers_on_modifier",
+		{ mode: "boolean" },
+	),
```
[the remark this answers is not in the document]
Comment (TrevorSmith-Wellsky) [reply packages/local-db/src/schema/schema.ts:240]: Fixed — added migration `0039_add_show_workspace_numbers_on_modifier.sql` in 8896b8729.

**Comment by tvrmsmith** [comment]
I will get back to this PR shortly.
